### ISSUE TITLE
FillMaskPipeline very slow when provided with a large `targets`

### ISSUE THREAD

**Issue by EtaoinWu** [huggingface/transformers#12099]
## Environment info

- `transformers` version: 4.6.1
- Platform: Linux-5.4.0-67-generic-x86_64-with-glibc2.10
- Python version: 3.8.5
- PyTorch version (GPU?): 1.8.1 (False)
- Tensorflow version (GPU?): N/A
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help
@LysandreJik @Narsil 

## Information

The model I am using: `ethanyt/guwenbert-base`, with a `RoBERTa` model and a `BertTokenizerFast` tokenizer.

## To reproduce

1. Initialize a `fill-mask` pipeline with the model and the tokenizer mentioned above
2. Call it with any sentence and a large `targets` (with a length of ~10k single words)

## Problem

The call would be much slower than a similar call without a `targets` argument. A call without a `targets` argument costs ~0.1s, while a call with a `targets` argument costs ~0.3s.

The following code is present in `src/transformers/pipelines/fill_mask.py`:

```python
class FillMaskPipeline(Pipeline):
    # ...
    def __call__(self, *args, targets=None, top_k: Optional[int] = None, **kwargs):
        # ...
        if targets is not None:
            # ...
            targets_proc = []
            for target in targets:
                target_enc = self.tokenizer.tokenize(target)
                # ...
                targets_proc.append(target_enc[0])
```

This function iterates through targets, rather than sending it directly to `tokenize`, which does not utilize the batch processing optimization of `TokenizerFast`s, hence the slow speed.

**Comment by EtaoinWu** [comment]
An example code can be found [in this colab example](https://colab.research.google.com/gist/EtaoinWu/0cf5b37882bd18bcc554d3da717a3974/fillmaskpipeline-test.ipynb). On the default Google machine that I wrote this notebook on, the version with a `targets` argument slows down significantly (100ish ms to 600ish ms).

> Caveat: when a target is going to be very long (like 20 tokens) with 10k targets, the resulting array will be 20 x 10k for ids, that can pile up quite fast memory usage. In that context, it could be much slower to pass everything at once. We need to benchmark that too.

The current behavior of `FillMaskPipeline` is that when a multi-token string is passed, only the first token is used. I doubt anyone would actually need this, because if someone want to choose a token from a subset of the vocabulary to fill into a mask, they usually know the subset exactly. Deliberately passing multi-token strings into `FillMaskPipeline` (and expecting it to tokenize them and drop all-but-first tokens) does not make much sense.

## Another discovery

When coding my example, I just discovered the bottleneck of the performance problem. When provided with a `targets` argument, `FillMaskPipeline` ignores its `top_k` parameter, which means that it has to output a whole list proportional to `len(targets)`, and that's the bottleneck (at least in my test). The code example above actually respects `top_k` parameter when a `targets` is present, hence much faster when constructing the return value. After the optimization, the code costs 200ish milliseconds.

**Comment by Narsil** [comment]
> An example code can be found [in this colab example](https://colab.research.google.com/gist/EtaoinWu/0cf5b37882bd18bcc554d3da717a3974/fillmaskpipeline-test.ipynb). On the default Google machine that I wrote this notebook on, the version with a `targets` argument slows down significantly (100ish ms to 600ish ms).

Thanks, this will help ! 
> 
> > Caveat: when a target is going to be very long (like 20 tokens) with 10k targets, the resulting array will be 20 x 10k for ids, that can pile up quite fast memory usage. In that context, it could be much slower to pass everything at once. We need to benchmark that too.
> 
> The current behavior of `FillMaskPipeline` is that when a multi-token string is passed, only the first token is used. I doubt anyone would actually need this, because if someone want to choose a token from a subset of the vocabulary to fill into a mask, they usually know the subset exactly. Deliberately passing multi-token strings into `FillMaskPipeline` (and expecting it to tokenize them and drop all-but-first tokens) does not make much sense.

As a maintainer of a live product, I can tell you not everyone is aware of what happens behind a pipeline (and it is exactly why they exist, so we can abstract away all nitty gritty details of transformers). So it will happen that some users will try out those examples and be surprised at slowness. 
It's something that `pipelines` should try to address if possible. 

> ### Another discovery
> 
> When coding my example, I just discovered the bottleneck of the performance problem. When provided with a `targets` argument, `FillMaskPipeline` ignores its `top_k` parameter, which means that it has to output a whole list proportional to `len(targets)`, and that's the bottleneck (at least in my test). The code example above actually respects `top_k` parameter when a `targets` is present, hence much faster when constructing the return value. After the optimization, the code costs 200ish milliseconds.

Ok, I think I remember the `targets` being added and the decision was that if `top_k` > `len(targets)` we were not obliged of honoring `top_k` because it wouldn't make any sense. `top_k` < `len(targets)` should be honored though.

**Comment by Narsil** [comment]
I was able to reproduce and optimize away most of the performance, now any example should run at roughly the same speed.

Slowdown will happen when you miss the vocabulary, but the warnings should help users figure it out.

**Comment by EtaoinWu** [comment]
Thanks a lot. As a background, I found the issue when reproducing the following paper:

> Deng, Liming, et al. "An Iterative Polishing Framework Based on Quality Aware Masked Language Model for Chinese Poetry Generation." _Proceedings of the AAAI Conference on Artificial Intelligence_. Vol. 34. No. 05. 2020.

which involves calling `FillMaskPipeline` iteratively 10 times at most for each API call, which depending on the input, may or may not have the `targets` parameter. The time difference in the two types of API calls made me find this issue.
